lib/nvme/nvme_qpair.c

```diff
@@ -611,28 +635,108 @@ nvme_qpair_construct(struct spdk_nvme_qpair *qpair, uint16_t id,
 }
 
 static void
-nvme_admin_qpair_abort_aers(struct spdk_nvme_qpair *qpair)
+nvme_admin_qpair_abort_aers_and_abort_cmd(struct spdk_nvme_qpair *qpair)
 {
 	struct nvme_tracker	*tr;
 
-	tr = LIST_FIRST(&qpair->outstanding_tr);
+	tr = TAILQ_FIRST(&qpair->outstanding_tr);
 	while (tr != NULL) {
 		assert(tr->req != NULL);
-		if (tr->req->cmd.opc == SPDK_NVME_OPC_ASYNC_EVENT_REQUEST) {
+		if (tr->req->cmd.opc == SPDK_NVME_OPC_ASYNC_EVENT_REQUEST || tr->req->cmd.opc == SPDK_NVME_OPC_ABORT) {
 			nvme_qpair_manual_complete_tracker(qpair, tr,
 							   SPDK_NVME_SCT_GENERIC, SPDK_NVME_SC_ABORTED_SQ_DELETION, 0,
 							   false);
-			tr = LIST_FIRST(&qpair->outstanding_tr);
+			tr = TAILQ_FIRST(&qpair->outstanding_tr);
 		} else {
-			tr = LIST_NEXT(tr, list);
+			tr = TAILQ_NEXT(tr, tq_list);
+		}
+	}
+}
+
+int
+nvme_qpair_check_timeout(struct spdk_nvme_qpair* qpair)
+{
+	int rc;
+	uint64_t t02;
+	struct nvme_tracker *tr;
+	struct nvme_request *req;
+	pthread_t tid;
+	pthread_attr_t tattr;
+
+	if (!TAILQ_EMPTY(&qpair->outstanding_tr)) {
+
+		/*
+		 * qpair could be either for normal i/o or for admin command. If qpair is admin
+		 * and request is SPDK_NVME_OPC_ASYNC_EVENT_REQUEST, skip to next previous.
+		 */
+
+		tr=TAILQ_LAST(&qpair->outstanding_tr, nvme_outstanding_tr_head);
+		while (tr->req->cmd.opc == SPDK_NVME_OPC_ASYNC_EVENT_REQUEST) {
+
+			/* qpair is for admin request */
+
+			tr = TAILQ_PREV(tr, nvme_outstanding_tr_head, tq_list);
+			if(!tr) {
+				/* 
+			 	 * All request were AER
+			 	 */
+				return 0;
+			}
+		}
+
+		req = tr->req;
+		t02 = rte_get_timer_cycles();
+		if (req->t0 <= t02){
+
+			if (!req->abort_count) {
+				/* 
+			 	 * Request has timed out. No previous Abort attempted for this request.
+				 * This could be i/o or admin request. Abort request and increase timeout.
+			 	 */
+				pthread_mutex_lock(&qpair->ctrlr->ctrlr_lock);
+				req->t0 = rte_get_timer_cycles() + qpair->ctrlr->hz * NVME_IO_TIMEOUT;
+				req->abort_count++;
+
+				qpair->ctrlr->abort_in_progress++;
+				pthread_attr_init(&tattr);
+				pthread_attr_setdetachstate(&tattr,PTHREAD_CREATE_DETACHED);
+				if ((rc=pthread_create(&tid, &tattr, abort_completions, qpair)) != 0) {
+					qpair->ctrlr->abort_in_progress--;
+					pthread_mutex_unlock(&qpair->ctrlr->ctrlr_lock);
+					nvme_printf(qpair->ctrlr, "error: pthread_create, rc =%d\n", rc);
+      					return rc;
+    				}
+
```

Comment:
If the admin queue is full, this will queue up the abort in software. This strategy would need to reserve a slot in the admin queue so that it can guarantee the command gets submitted to the device immediately.